### PR TITLE
support multiple ranges in attribute partial update

### PR DESCRIPTION
**Pull request by Pessimistress**
For #1576 

#### Change List
- Support invalidating a partial range of an attribute
- Unit tests

### CODE REVIEW

File: modules/core/src/lib/attribute.js [hunk 1]
```diff
@@ -176,47 +187,57 @@ export default class Attribute extends BaseAttribute {
         this.buffer.reallocate(this.value.byteLength);
       }
 
-      state.needsUpdate = true;
+      if (state.updateRanges !== range.FULL) {
+        this.value.set(oldValue);
+        // TODO - copy on the GPU
+        this.buffer.subData(oldValue);
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Maybe some comment here. I would naively expect we would need to specify an offset if range is not `FULL`?

File: modules/core/src/lib/attribute.js [hunk 2]
```diff
@@ -176,47 +187,57 @@ export default class Attribute extends BaseAttribute {
...
     let updated = true;
     if (update) {
       // Custom updater - typically for non-instanced layers
-      update.call(context, this, {data, startRow, endRow, props, numInstances, bufferLayout});
+      for (const [startRow, endRow] of updateRanges) {
+        update.call(context, this, {data, startRow, endRow, props, numInstances, bufferLayout});
+      }
       if (this.constant || !this.buffer || this.buffer.byteLength < this.value.byteLength) {
         // Full update
         this.update({
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: This branch calls a helper method but the other branch does the update inline. Maybe align (e.g. use helper methods in both branches, or have the full update just prepare a range that covers all and fall through)?

File: modules/core/src/utils/range.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    if (range0[1] < range[0]) {
+      // the current range is to the left of the new range
+      newRangeCollection.push(range0);
+      insertPosition = i + 1;
+    } else if (range0[0] > range[1]) {
+      // the current range is to the right of the new range
+      newRangeCollection.push(range0);
+    } else {
+      range = [Math.min(range0[0], range[0]), Math.max(range0[1], range[1])];
```
Comment (ibgreen) [inline modules/core/src/utils/range.js]: Just my 2 cents: Managing overlapping ranges seems like the user's problem. I.e. I would have avoided adding this complexity/additional code size here (to keep the "monolith" from growing too fast). 

An optional, tree-shakeable `Range` class could have been an alternative, either here or in math.gl

File: modules/core/src/utils/range.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  // Validate the input range
+  if (range[0] < 0) {
+    range[0] = 0;
+  }
+  if (range[0] >= range[1]) {
+    return rangeCollection;
+  }
+
+  const newRangeCollection = [];
```
Comment (ibgreen) [inline modules/core/src/utils/range.js]: Total nit: Not sure about the "collection" term. Has other connotations to me (base class of containers). I assume you want to emphasize that it has some additional semantics over a list. I think "rangeList" would have been fine.

File: modules/core/src/lib/attribute.js (hunk 2 above)
Comment (Pessimistress) [reply modules/core/src/lib/attribute.js]: `update` is a method of the `BaseAttribute` class that uploads the attribute value to the GPU. It's certainly confusing given the various things called `update` here.

**Pull request merged**